cg.go

```diff
@@ -0,0 +1,317 @@
+	// AngleRestartThreshold, that is, if
+	//
+	//   ∇f_k·∇f_{k+1} / (|∇f_k| |∇f_{k+1}|) < AngleRestartThreshold,
+	//
+	// the descent direction is the same as the negative of the gradient.
+	// A value of AngleRestartThreshold closer to -1 will reduce the number of
+	// restarts.
+	// If AngleRestartThreshold is 0, it will be set to -0.9.
+	// CG will panic if AngleRestartThreshold is not in the interval [-1, 0].
+	AngleRestartThreshold float64
+
+	linesearch *Linesearch
+
+	restartAfter    int
+	iterFromRestart int
+
+	dirPrev      []float64
+	gradPrev     []float64
+	gradPrevNorm float64
+}
+
+func (cg *CG) Init(loc Location, funcInfo *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	if cg.IterationRestartFactor < 0 {
+		panic("cg: IterationRestartFactor is negative")
+	}
+	if cg.AngleRestartThreshold < -1 || cg.AngleRestartThreshold > 0 {
+		panic("cg: AngleRestartThreshold not in [-1, 0]")
+	}
+
+	if cg.LinesearchMethod == nil {
+		cg.LinesearchMethod = &Bisection{GradConst: 0.1}
+	}
+	if cg.Variant == nil {
+		cg.Variant = &HestenesStiefel{}
+	}
+	if cg.InitialStep == nil {
+		cg.InitialStep = &FirstOrderStepSize{}
+	}
+
+	if cg.IterationRestartFactor == 0 {
+		cg.IterationRestartFactor = iterationRestartFactor
+	}
+	if cg.AngleRestartThreshold == 0 {
+		cg.AngleRestartThreshold = angleRestartThreshold
+	}
+
+	if cg.linesearch == nil {
+		cg.linesearch = &Linesearch{}
+	}
+	cg.linesearch.Method = cg.LinesearchMethod
+	cg.linesearch.NextDirectioner = cg
+
+	return cg.linesearch.Init(loc, funcInfo, xNext)
+}
+
+func (cg *CG) Iterate(loc Location, xNext []float64) (EvaluationType, IterationType, error) {
+	return cg.linesearch.Iterate(loc, xNext)
+}
+
+func (cg *CG) InitDirection(loc Location, direction []float64) (stepSize float64) {
+	dim := len(loc.X)
+
+	cg.restartAfter = int(math.Ceil(cg.IterationRestartFactor * float64(dim)))
+	cg.iterFromRestart = 0
+
+	// The initial direction is always the negative gradient.
+	copy(direction, loc.Gradient)
+	floats.Scale(-1, direction)
+
+	cg.dirPrev = resize(cg.dirPrev, dim)
+	copy(cg.dirPrev, direction)
+	cg.gradPrev = resize(cg.gradPrev, dim)
+	copy(cg.gradPrev, loc.Gradient)
+	cg.gradPrevNorm = floats.Norm(loc.Gradient, 2)
+
+	cg.Variant.Init(loc)
+	return cg.InitialStep.Init(loc, direction)
+}
+
+func (cg *CG) NextDirection(loc Location, direction []float64) (stepSize float64) {
+	copy(direction, loc.Gradient)
+	floats.Scale(-1, direction)
+
+	cg.iterFromRestart++
+	var restart bool
+	// Check if the method should be restarted because too many iterations have
+	// been taken without a restart.
+	if cg.iterFromRestart == cg.restartAfter {
+		restart = true
+	}
+	// Check if the method should be restarted because the angle between
+	// the last two gradients is too large.
+	gDot := floats.Dot(loc.Gradient, cg.gradPrev)
+	gNorm := floats.Norm(loc.Gradient, 2)
+	cos := math.Max(-1, gDot/(gNorm*cg.gradPrevNorm))
```

Comment:
What does the Max do? It seems to me like the only effect is has is to allow AngleRestartThreshold == -1 to act as "never restart because of the angle" even when the gradients are really poorly scaled. If the line below this was "<=" instead of less than, the max would have no effect, because cg.AngleRestartThreshold >= -1.